File: backend/internal/services/handicap.go
// services/handicap.go
// Pure handicap math + the cross-cutting score-recalculation routine.
//
// These live in the services package (not in any specific service struct)
// because both EventService and the future ScoreService need them: changing an
// event's handicap allowance retriggers the calc on every score in every round
// of that event. Keeping the helpers here avoids circular imports between
// services that depend on each other for a stateless arithmetic concern.
//
// PR #4 will move RecalculateEventScores onto a ScoreService; the pure
// arithmetic helpers (HandicapStrokes, EffectiveCourseHandicap) likely stay
// in this file regardless.
package services

import (
	"context"
	"fmt"
	"math"
	"sort"

	"github.com/google/uuid"
	"github.com/trentd187/golf-league/internal/models"
	"gorm.io/gorm"
)

// HandicapStrokes returns the strokes a player with the given course handicap
// receives (positive) or gives back (negative) on a hole identified by its
// normalized stroke_index rank. Net is always gross - strokes, so a negative
// return (a plus handicap) raises the player's net on that hole.
//
// strokeIndex must be a rank within the played set (1 = hardest hole, holeCount =
// easiest). holeCount is the number of holes being played (9 or 18).
//
// USGA allocation rule generalized: a player with handicap 5 on a 9-hole round
// gets one stroke on the 5 hardest holes (normalized SI 1–5). A player with
// handicap 20 on an 18-hole round gets two strokes on the two hardest holes and
// one stroke on the remaining 16. A PLUS handicap (negative, better than scratch)
// gives strokes back starting at the EASIEST hole (highest SI) and working down —
// the mirror of receiving. This must stay identical to the client
// holeHandicapStrokes so client-computed net matches the server.
func HandicapStrokes(courseHandicap, strokeIndex, holeCount int) int {
	if courseHandicap == 0 || strokeIndex <= 0 || holeCount <= 0 {
		return 0
	}
	if courseHandicap > 0 {
		full := courseHandicap / holeCount      // complete passes over all holes
		remainder := courseHandicap % holeCount // extra strokes from SI 1 upward
		strokes := full
		if strokeIndex <= remainder {
			strokes++
		}
		return strokes
	}
	// Plus handicap: give `give` strokes back, allocated from the easiest hole down.
	give := -courseHandicap
	full := give / holeCount
	remainder := give % holeCount // extra given strokes from the easiest hole down
	strokes := full
	if strokeIndex > holeCount-remainder {
		strokes++
	}
	return -strokes
}

// Course-handicap sanity bounds. WHS caps a Course Handicap at 54; a plus handicap
// (negative, better than scratch) rarely exceeds +10. These reject fat-fingered
// entries (e.g. "+99") that would otherwise allocate dozens of phantom strokes.
const (
	MinCourseHandicap = -10
	MaxCourseHandicap = 54
)

// validateCourseHandicap returns a *ValidationError when a course handicap is
// outside the sane WHS range. Whether an unset (nil) handicap is allowed is the
// caller's concern — this only bounds a value that is present.
func validateCourseHandicap(h int) error {
	if h < MinCourseHandicap || h > MaxCourseHandicap {
		return &ValidationError{
			Field:   "course_handicap",
			Message: fmt.Sprintf("course_handicap must be between %d and %d", MinCourseHandicap, MaxCourseHandicap),
		}
	}
	return nil
}

// NormalizeStrokeIndexes returns a map from hole_number → normalized rank (1 = hardest).
// Holes are ranked by ascending stroke_index so handicap allocation works correctly
// when playing a subset of holes (e.g. front or back 9 of an 18-hole course whose
// raw stroke indexes span 1–18 with gaps).
func NormalizeStrokeIndexes(holes []models.Hole) map[int]int {
	type entry struct{ holeNumber, si int }
	entries := make([]entry, 0, len(holes))
	for _, h := range holes {
		entries = append(entries, entry{h.HoleNumber, h.StrokeIndex})
	}
	sort.Slice(entries, func(i, j int) bool { return entries[i].si < entries[j].si })
	m := make(map[int]int, len(entries))
	for rank, e := range entries {
		m[e.holeNumber] = rank + 1
	}
	return m
}

// filterPlayedHoles returns the subset of holes being played for a round.
// nil selection → all holes; "front" → holes 1–9; "back" → holes 10–18.
func filterPlayedHoles(holes []models.Hole, sel *string) []models.Hole {
	if sel == nil {
		return holes
	}
	result := make([]models.Hole, 0, 9)
	for _, h := range holes {
		if *sel == "front" && h.HoleNumber <= 9 {
			result = append(result, h)
		} else if *sel == "back" && h.HoleNumber > 9 {
			result = append(result, h)
		}
	}
	return result
}

// EffectiveCourseHandicap applies the event's handicap allowance percentage to
// a player's raw course handicap.
//
//	allowance = nil  → no allowance set; full handicap.
//	allowance = 90.0 → effective = floor(raw * 0.90).
//
// floor() is USGA convention so the result is always an integer.
func EffectiveCourseHandicap(courseHandicap int, allowance *float64) int {
	if allowance == nil {
		return courseHandicap
	}
	return int(math.Floor(float64(courseHandicap) * (*allowance) / 100.0))
}

// RecalculateEventScores recomputes net_score for every scored hole across all
// rounds in an event. Triggered when an event's handicap_allowance changes.
//
// Processes per-round so each round's nine_hole_selection can be used to
// normalize stroke indexes before applying HandicapStrokes.
//
// All-or-nothing. The whole recalculation runs in ONE transaction: it rewrites net_score for
// every scored hole in the event, and a failure partway through used to leave half the event
// on the old allowance and half on the new — a silently wrong leaderboard that nothing
// repairs, because the allowance itself has already been saved and a retry sees no work to do.
// Either every score reflects the new allowance or none does.
func RecalculateEventScores(ctx context.Context, db *gorm.DB, eventID uuid.UUID, allowance *float64) error {
	return db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
		var rounds []models.Round
		if err := tx.
			Preload("DefaultTee.Holes").
			Where("event_id = ?", eventID).
			Find(&rounds).Error; err != nil {
			return fmt.Errorf("load rounds for recalc: %w", err)
		}

		type scoreRow struct {
			ScoreID        uuid.UUID
			GrossScore     int
			HoleNumber     int
			CourseHandicap *int
		}

		for _, round := range rounds {
			played := filterPlayedHoles(round.DefaultTee.Holes, round.NineHoleSelection)
			if len(played) == 0 {
				continue
			}
			siMap := NormalizeStrokeIndexes(played)
			holeCount := len(played)

			var rows []scoreRow
			if err := tx.Table("scores s").
				Select("s.id as score_id, s.gross_score, s.hole_number, rp.course_handicap").
				Joins("JOIN round_players rp ON rp.id = s.round_player_id").
				Where("rp.round_id = ?", round.ID).
				Scan(&rows).Error; err != nil {
				return fmt.Errorf("load scores for round %s: %w", round.ID, err)
			}

			for _, row := range rows {
				raw := 0
				if row.CourseHandicap != nil {
					raw = *row.CourseHandicap
				}
				eff := EffectiveCourseHandicap(raw, allowance)
				netScore := row.GrossScore - HandicapStrokes(eff, siMap[row.HoleNumber], holeCount)

				if err := tx.Model(&models.Score{}).
					Where("id = ?", row.ScoreID).
					Update("net_score", netScore).Error; err != nil {
					return fmt.Errorf("update score %s: %w", row.ScoreID, err)
				}
			}
		}
		return nil
	})
}

File: mobile/utils/handicap.ts
// utils/handicap.ts
// Pure handicap-stroke allocation shared by every team format that needs net
// scores derived client-side (Las Vegas, Best Ball, …). Kept React-free so it can
// be unit-tested in isolation and reused without pulling in any format's helpers.
// Mirrors the backend allocation so client-computed net matches the server exactly.

import type { ScorecardHole } from "@/types/scorecard";

// normalizeStrokeIndexes ranks holes by ascending stroke_index (1 = hardest) so
// handicap strokes allocate correctly even when playing a 9-hole subset. Mirrors
// the backend NormalizeStrokeIndexes so client-computed net matches the server.
export function normalizeStrokeIndexes(holes: ScorecardHole[]): Record<number, number> {
  const sorted = [...holes].sort((a, b) => a.stroke_index - b.stroke_index);
  const map: Record<number, number> = {};
  sorted.forEach((h, i) => {
    map[h.hole_number] = i + 1;
  });
  return map;
}

// holeHandicapStrokes returns the strokes a player receives (positive) or gives
// back (negative) on a hole given their effective handicap, the hole's normalized
// stroke-index rank, and the hole count. Net is always `gross - strokes`, so a
// negative return (plus handicap) raises the player's net on that hole.
// Mirrors the backend HandicapStrokes allocation rule.
//
// A POSITIVE handicap receives strokes on the hardest holes first (normalizedSI 1 up).
// A PLUS handicap (negative, better than scratch) gives strokes back starting at the
// EASIEST hole (highest normalizedSI) and working down — the USGA convention, the
// exact mirror of receiving. A negative value here means "add a stroke to gross".
export function holeHandicapStrokes(effHandicap: number, normalizedSI: number, holeCount: number): number {
  if (effHandicap === 0 || normalizedSI <= 0 || holeCount <= 0) return 0;
  if (effHandicap > 0) {
    const full = Math.floor(effHandicap / holeCount);
    const remainder = effHandicap % holeCount;
    return full + (normalizedSI <= remainder ? 1 : 0);
  }
  // Plus handicap: give `give` strokes back, allocated from the easiest hole down.
  const give = -effHandicap;
  const full = Math.floor(give / holeCount);
  const remainder = give % holeCount;
  const given = full + (normalizedSI > holeCount - remainder ? 1 : 0);
  return given === 0 ? 0 : -given; // avoid returning -0 for holes with no given stroke
}

// parseCourseHandicapInput converts a raw text input to the stored course_handicap
// value, using golf notation: a leading "+" marks a PLUS handicap (better than
// scratch) and is stored as a negative (e.g. "+2" → -2); plain digits are a regular
// handicap ("12" → 12); "0" is scratch. Empty/whitespace or anything else (a bare
// "-", a decimal, stray characters) → null so a typo never blocks the save — the
// caller shows an error and the organizer can correct it.
export function parseCourseHandicapInput(text: string): number | null {
  const trimmed = text.trim();
  if (trimmed === "") return null;
  if (!/^\+?\d+$/.test(trimmed)) return null;
  const digits = Number(trimmed.replace("+", ""));
  if (!Number.isInteger(digits)) return null;
  if (digits === 0) return 0; // avoid producing -0 for "+0"
  // A leading "+" means a plus handicap, stored as negative.
  return trimmed.startsWith("+") ? -digits : digits;
}

// formatCourseHandicap renders a stored course_handicap for display in golf notation:
// a plus handicap (negative) shows with a "+" prefix (-2 → "+2"), scratch is "0",
// and a regular handicap is its plain number (12 → "12").
export function formatCourseHandicap(stored: number): string {
  if (stored < 0) return `+${-stored}`;
  return String(stored);
}

// formatCourseHandicapWithAllowance renders a player's course handicap for the C.H.
// header, appending the allowance-adjusted effective handicap when an allowance is
// active and actually changes the value (e.g. "+2 → +1"). Otherwise just the plain
// handicap ("+2", "12"). Empty string when the handicap is unset. This is the header
// home for the raw→effective conversion that used to clutter every per-hole box.
export function formatCourseHandicapWithAllowance(
  courseHandicap: number | null | undefined,
  effective: number | null | undefined,
  allowanceActive: boolean,
): string {
  if (courseHandicap == null) return "";
  const base = formatCourseHandicap(courseHandicap);
  if (allowanceActive && effective != null && effective !== courseHandicap) {
    return `${base} → ${formatCourseHandicap(effective)}`;
  }
  return base;
}

// formatHoleStrokeAdjustment renders a hole's handicap strokes as the signed change
// applied to gross to get net (i.e. `-strokes`): a received stroke lowers the score
// ("−1"), a plus-handicap given stroke raises it ("+1"). Uses a true minus sign (−)
// for clarity. Callers only render this when strokes !== 0.
export function formatHoleStrokeAdjustment(strokes: number): string {
  if (strokes > 0) return `−${strokes}`; // received: subtracted from gross
  return `+${-strokes}`; // given (plus handicap): added to gross
}
